aptos-move/framework/aptos-token/sources/token_event_utils.move

```diff
@@ -0,0 +1,318 @@
+    }
+
+    /// Event emitted when collection uri is mutated
+    struct CollectionUriMutateEvent has drop, store {
+        creator_addr: address,
+        collection_name: String,
+        /// new uri
+        uri: String,
+    }
+
+    /// Event emitted when the collection maximum is mutated
+    struct CollectionMaxiumMutateEvent has drop, store {
+        creator_addr: address,
+        collection_name: String,
+        /// new maximum
+        maximum: u64,
+    }
+
+    //
+    // Token transfer related events
+    //
+
+    /// Event emitted when an user opt-in the direct transfer
+    struct OptInTransferEvent has drop, store {
+        /// True if the user opt in, false if the user opt-out
+        opt_in: bool
+    }
+
+    //
+    // Token mutation events
+    //
+
+    /// Event emitted when the tokendata uri mutates
+    struct UriMutationEvent has drop, store {
+        creator: address,
+        collection: String,
+        token: String,
+        /// new URI
+        new_uri: String,
+    }
+
+    /// Event emitted when mutating the default the token properties stored at tokendata
+    struct DefaultPropertyMutateEvent has drop, store {
+        creator: address,
+        collection: String,
+        token: String,
+        new_keys: vector<String>,
+        new_values: vector<vector<u8>>,
+        new_types: vector<String>,
+    }
+
+    /// Event emitted when the tokendata description is mutated
+    struct DescriptionMutateEvent has drop, store {
+        creator: address,
+        collection: String,
+        token: String,
+        description: String,
+    }
+
+    /// Event emitted when the token royalty is mutated
+    struct RoyaltyMutateEvent has drop, store {
+        creator: address,
+        collection: String,
+        token: String,
+        royalty_numerator: u64,
+        royalty_denominator: u64,
+        royalty_payee_addr: address,
+    }
+
+    /// Event emitted when the token maximum is mutated
+    struct MaxiumMutateEvent has drop, store {
+        creator: address,
+        collection: String,
+        token: String,
+        maximum: u64,
+    }
+
+    struct TokenEventStore has key {
+        /// collection mutation events
+        collection_uri_mutate_events: EventHandle<CollectionUriMutateEvent>,
+        collection_maximum_mutate_events: EventHandle<CollectionMaxiumMutateEvent>,
+        collection_description_mutate_events: EventHandle<CollectionDescriptionMutateEvent>,
+        /// token transfer opt-in event
+        opt_in_events: EventHandle<OptInTransferEvent>,
+        /// token mutation events
+        uri_mutate_events: EventHandle<UriMutationEvent>,
+        default_property_mutate_events: EventHandle<DefaultPropertyMutateEvent>,
+        description_mutate_events: EventHandle<DescriptionMutateEvent>,
+        royalty_mutate_events: EventHandle<RoyaltyMutateEvent>,
+        maximum_mutate_events: EventHandle<MaxiumMutateEvent>,
+        /// This is for adding new events in future
```

Comment:
There has been some discussion in the past but how would this work when we need to add a new event? Can this lead to some sort of deeply nested state when we do a few rounds of new events?

Reply:
I added an option<Any> in the token event store where we can add new struct for new events

Reply:
Alright makes sense. We should revamp events in Q1. It's definitely a pain currently to use and extend :)